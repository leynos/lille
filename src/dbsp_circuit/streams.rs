//! DBSP dataflow stream construction for spatial simulation.
//!
//! This module defines helper functions for building the dataflow streams used
//! by `DbspCircuit` to process the game world. These streams implement:
//!
//! - Block aggregation to track the highest block at each grid cell
//! - Floor height calculation with optional slopes
//! - Velocity updates applying gravity
//! - Position integration based on velocities
//! - Joining positions with floor height for collision queries
//!
//! Each function returns a new [`Stream`] that can be composed into the overall
//! circuit.

use dbsp::{operator::Max, typed_batch::OrdZSet, RootCircuit, Stream};
use ordered_float::OrderedFloat;

use crate::components::{Block, BlockSlope};
use crate::{BLOCK_CENTRE_OFFSET, BLOCK_TOP_OFFSET, GRAVITY_PULL};

use super::{FloorHeightAt, HighestBlockAt, Position, Velocity};

/// Returns a stream pairing each grid cell with its highest block and id.
///
/// The function aggregates incoming [`Block`] records by `(x, y)` to find the
/// maximum `z` value at each coordinate. The output preserves the originating
/// block id so that subsequent joins can access slope information.
///
/// # Examples
///
/// ```rust,no_run
/// # fn main() -> Result<(), dbsp::Error> {
/// # use lille::prelude::*;
/// # use dbsp::{RootCircuit, typed_batch::OrdZSet};
/// # use lille::dbsp_circuit::highest_block_pair;
/// RootCircuit::build(|circuit| {
///     let (stream, _handle) = circuit.add_input_zset::<Block>();
///     let _highest = lille::dbsp_circuit::highest_block_pair(&stream);
///     Ok(())
/// })?;
/// # Ok::<(), dbsp::Error>(())
/// # }
/// ```
pub fn highest_block_pair(
    blocks: &Stream<RootCircuit, OrdZSet<Block>>,
) -> Stream<RootCircuit, OrdZSet<(HighestBlockAt, i64)>> {
    blocks
        .map_index(|b| ((b.x, b.y), (b.z, b.id)))
        .aggregate(Max)
        .map(|((x, y), (z, id))| {
            (
                HighestBlockAt {
                    x: *x,
                    y: *y,
                    z: *z,
                },
                *id,
            )
        })
}

/// Derives the floor height for each block, optionally applying slopes.
///
/// The stream joins the highest block id at a grid cell with any matching
/// [`BlockSlope`] record. When slope data is present the returned
/// [`FloorHeightAt`] accounts for the block's gradient, producing a smooth
/// surface. Missing slope data falls back to a flat top.
pub(super) fn floor_height_stream(
    highest_pair: &Stream<RootCircuit, OrdZSet<(HighestBlockAt, i64)>>,
    slopes: &Stream<RootCircuit, OrdZSet<BlockSlope>>,
) -> Stream<RootCircuit, OrdZSet<FloorHeightAt>> {
    highest_pair
        .map_index(|(hb, id)| (*id, (hb.x, hb.y, hb.z)))
        .outer_join(
            &slopes.map_index(|bs| (bs.block_id, (bs.grad_x, bs.grad_y))),
            |_, &(x, y, z), &(gx, gy)| {
                let base = z as f64 + BLOCK_TOP_OFFSET;
                let gradient = BLOCK_CENTRE_OFFSET * (gx.into_inner() + gy.into_inner());
                Some(FloorHeightAt {
                    x,
                    y,
                    z: OrderedFloat(base + gradient),
                })
            },
            |_, &(x, y, z)| {
                Some(FloorHeightAt {
                    x,
                    y,
                    z: OrderedFloat(z as f64 + BLOCK_TOP_OFFSET),
                })
            },
            |_, _| None,
        )
        // Convert `Option<FloorHeightAt>` from the outer join, discarding
        // unmatched slope records. `flat_map` passes each item by reference,
        // so cloning is required to produce an owned `Option` for the
        // iterator.
        .flat_map(|fh| fh.clone())
}

/// Applies gravity to each velocity record.
///
/// This helper keeps the velocity update logic separate from entity position
/// updates. It adds [`GRAVITY_PULL`] to the incoming `vz` component and passes
/// through the remaining fields unchanged.
pub(super) fn new_velocity_stream(
    velocities: &Stream<RootCircuit, OrdZSet<Velocity>>,
) -> Stream<RootCircuit, OrdZSet<Velocity>> {
    velocities.map(|v| Velocity {
        entity: v.entity,
        vx: v.vx,
        vy: v.vy,
        vz: OrderedFloat(v.vz.into_inner() + GRAVITY_PULL),
    })
}

/// Integrates positions with updated velocities.
///
/// The input streams are joined by `entity`, producing a new [`Position`]
/// translated by the entity's velocity components. The function performs a
/// simple Euler integration suitable for the small time step used in the game
/// loop.
pub(super) fn new_position_stream(
    positions: &Stream<RootCircuit, OrdZSet<Position>>,
    new_vel: &Stream<RootCircuit, OrdZSet<Velocity>>,
) -> Stream<RootCircuit, OrdZSet<Position>> {
    positions.map_index(|p| (p.entity, p.clone())).join(
        &new_vel.map_index(|v| (v.entity, v.clone())),
        |_, p, v| Position {
            entity: p.entity,
            x: OrderedFloat(p.x.into_inner() + v.vx.into_inner()),
            y: OrderedFloat(p.y.into_inner() + v.vy.into_inner()),
            z: OrderedFloat(p.z.into_inner() + v.vz.into_inner()),
        },
    )
}

#[derive(
    rkyv::Archive,
    rkyv::Serialize,
    rkyv::Deserialize,
    Clone,
    Debug,
    PartialEq,
    Eq,
    Hash,
    PartialOrd,
    Ord,
    Default,
    size_of::SizeOf,
)]
#[archive_attr(derive(Ord, PartialOrd, Eq, PartialEq, Hash))]
/// Pairs an entity's position with the floor height at its grid location.
///
/// This struct is the result of joining the continuous [`Position`] stream with
/// the discrete [`FloorHeightAt`] grid. It is primarily consumed by higher-level
/// physics logic for tasks such as collision detection or standing vs. falling
/// checks.
///
/// * `position` - The entity's current position in world coordinates
/// * `z_floor` - The computed floor height at the position's grid cell
pub struct PositionFloor {
    pub position: Position,
    pub z_floor: OrderedFloat<f64>,
}

/// Joins each `Position` with the corresponding floor height.
///
/// Positions are discretised to grid coordinates by flooring their `x` and `y`
/// values. Those indices look up a [`FloorHeightAt`] record to produce a
/// [`PositionFloor`] stream suitable for higher-level physics logic.
pub(super) fn position_floor_stream(
    positions: &Stream<RootCircuit, OrdZSet<Position>>,
    floor_height: &Stream<RootCircuit, OrdZSet<FloorHeightAt>>,
) -> Stream<RootCircuit, OrdZSet<PositionFloor>> {
    positions
        .map_index(|p| {
            (
                (
                    p.x.into_inner().floor() as i32,
                    p.y.into_inner().floor() as i32,
                ),
                p.clone(),
            )
        })
        .join(
            &floor_height.map_index(|fh| ((fh.x, fh.y), fh.z)),
            |_idx, pos, &z_floor| PositionFloor {
                position: pos.clone(),
                z_floor,
            },
        )
}

/// Computes new positions and velocities for entities standing on the ground.
///
/// Standing entities move according to their horizontal velocity components and
/// snap to the floor height at their new `(x, y)` coordinates. The vertical
/// velocity is reset to zero to keep entities grounded.
pub(super) fn standing_motion_stream(
    standing: &Stream<RootCircuit, OrdZSet<PositionFloor>>,
    floor_height: &Stream<RootCircuit, OrdZSet<FloorHeightAt>>,
    velocities: &Stream<RootCircuit, OrdZSet<Velocity>>,
) -> (
    Stream<RootCircuit, OrdZSet<Position>>,
    Stream<RootCircuit, OrdZSet<Velocity>>,
) {
    let moved = standing
        .map_index(|pf| (pf.position.entity, pf.position.clone()))
        .join(
            &velocities.map_index(|v| (v.entity, v.clone())),
            |_, pos, vel| {
                let new_x = OrderedFloat(pos.x.into_inner() + vel.vx.into_inner());
                let new_y = OrderedFloat(pos.y.into_inner() + vel.vy.into_inner());
                (new_x, new_y, pos.entity, vel.vx, vel.vy)
            },
        );

    let indexed = moved.map_index(|(x, y, entity, vx, vy)| {
        (
            (x.into_inner().trunc() as i32, y.into_inner().trunc() as i32),
            (*entity, *x, *y, *vx, *vy),
        )
    });

    let with_floor = indexed.join(
        &floor_height.map_index(|fh| ((fh.x, fh.y), fh.z)),
        |_idx, &(entity, x, y, vx, vy), &z_floor| {
            (
                Position {
                    entity,
                    x,
                    y,
                    z: z_floor,
                },
                Velocity {
                    entity,
                    vx,
                    vy,
                    vz: OrderedFloat(0.0),
                },
            )
        },
    );

    let new_pos = with_floor.map(|(p, _)| p.clone());
    let new_vel = with_floor.map(|(_, v)| v.clone());
    (new_pos, new_vel)
}
