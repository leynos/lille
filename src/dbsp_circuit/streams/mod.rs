//! DBSP dataflow stream construction for spatial simulation.
//!
//! This module is split into domain-focused submodules and re-exports their
//! helpers for building the overall circuit.

pub(super) mod behaviour;
pub(super) mod floor;
pub(super) mod kinematics;

#[cfg(test)]
pub mod test_utils;

pub use behaviour::{apply_movement, fear_level_stream, movement_decision_stream};
pub use floor::{floor_height_stream, highest_block_pair};
pub use kinematics::{
    new_position_stream, new_velocity_stream, position_floor_stream, standing_motion_stream,
    PositionFloor,
};
