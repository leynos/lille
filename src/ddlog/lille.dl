typedef EntityID = signed<64>
typedef BlockID = signed<64>

// Continuous coordinate used for sub-block precision
typedef GCoord = float

import fp
import souffle_lib
import constants
import geometry
import entity_state
import physics

extern function vec_mag(x: GCoord, y: GCoord, z: GCoord): GCoord
extern function vec_normalize(x: GCoord, y: GCoord, z: GCoord): (GCoord, GCoord, GCoord)
