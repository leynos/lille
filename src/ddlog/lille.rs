/// Entry point for custom DDlog helpers.
///
/// The generated DDlog crate expects these functions to be in scope when it is
/// built. Re-export them so the main crate and tests can call the same
/// implementations that DDlog uses.
pub use types__physics::{sign, vec_mag, vec_normalize};
