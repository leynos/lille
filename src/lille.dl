typedef EntityID = signed64
typedef Coord = signed32
typedef Health = signed32
typedef UnitType = string

input relation Position(entity: EntityID, x: Coord, y: Coord)
input relation HealthRel(entity: EntityID, hp: Health)
input relation Unit(entity: EntityID, type: UnitType)
input relation Target(entity: EntityID, tx: Coord, ty: Coord)
input relation Fraidiness(entity: EntityID, factor: float)
input relation Meanness(entity: EntityID, factor: float)

output relation NewPosition(entity: EntityID, x: Coord, y: Coord)
output relation Damage(target: EntityID, amount: Health)
output relation Despawn(entity: EntityID)

