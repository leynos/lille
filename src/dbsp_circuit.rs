//! DBSP-based world inference engine.
//!
//! This module defines [`DbspCircuit`], the authoritative dataflow program for
//! Lille's game world. Callers feed [`Position`] and [`Block`] records into the
//! circuit, call [`DbspCircuit::step`], then read [`NewPosition`] and
//! [`HighestBlockAt`] outputs. Input collections persist across stepsâ€”invoke
//! [`DbspCircuit::clear_inputs`] after each frame to prevent stale data from
//! affecting subsequent computations.
//!
//! # Doctest prelude
//!
//! Examples in this module use the following imports:
//!
//! ```rust,no_run
//! # use lille::prelude::*;
//! ```

use anyhow::Error as AnyError;
use dbsp::{typed_batch::OrdZSet, CircuitHandle, OutputHandle, RootCircuit, ZSetHandle};

use crate::components::{Block, BlockSlope};
use crate::GRACE_DISTANCE;

mod streams;
mod types;
use streams::{
    apply_movement, fear_level_stream, floor_height_stream, movement_decision_stream,
    new_position_stream, new_velocity_stream, position_floor_stream, standing_motion_stream,
};
pub use streams::{highest_block_pair, PositionFloor};
pub use types::{
    FearLevel, FloorHeightAt, Force, HighestBlockAt, MovementDecision, NewPosition, NewVelocity,
    Position, Target, Velocity,
};

pub struct DbspCircuit {
    circuit: CircuitHandle,
    position_in: ZSetHandle<Position>,
    velocity_in: ZSetHandle<Velocity>,
    force_in: ZSetHandle<Force>,
    target_in: ZSetHandle<Target>,
    block_in: ZSetHandle<Block>,
    block_slope_in: ZSetHandle<BlockSlope>,
    new_position_out: OutputHandle<OrdZSet<NewPosition>>,
    new_velocity_out: OutputHandle<OrdZSet<NewVelocity>>,
    highest_block_out: OutputHandle<OrdZSet<HighestBlockAt>>,
    floor_height_out: OutputHandle<OrdZSet<FloorHeightAt>>,
    position_floor_out: OutputHandle<OrdZSet<PositionFloor>>,
}

struct BuildHandles {
    position_in: ZSetHandle<Position>,
    velocity_in: ZSetHandle<Velocity>,
    force_in: ZSetHandle<Force>,
    target_in: ZSetHandle<Target>,
    block_in: ZSetHandle<Block>,
    block_slope_in: ZSetHandle<BlockSlope>,
    new_position_out: OutputHandle<OrdZSet<NewPosition>>,
    new_velocity_out: OutputHandle<OrdZSet<NewVelocity>>,
    highest_block_out: OutputHandle<OrdZSet<HighestBlockAt>>,
    floor_height_out: OutputHandle<OrdZSet<FloorHeightAt>>,
    position_floor_out: OutputHandle<OrdZSet<PositionFloor>>,
}

impl DbspCircuit {
    /// Constructs a new `DbspCircuit` for simulating game world physics and environment state.
    ///
    /// Sets up a DBSP dataflow circuit with input handles for entity positions, velocities, and blocks.
    /// The circuit computes updated velocities by applying gravity, joins them with positions to
    /// produce new positions, and aggregates block data to determine the highest block at each
    /// `(x, y)` coordinate. Returns input and output handles for external interaction.
    ///
    /// # Returns
    ///
    /// A new `DbspCircuit` instance on success, or a DBSP error if circuit construction fails.
    ///
    /// # Examples
    ///
    /// ```rust,no_run
    /// # use lille::prelude::*;
    /// let circuit = DbspCircuit::new().expect("circuit construction failed");
    /// ```
    pub fn new() -> Result<Self, dbsp::Error> {
        let (circuit, handles) = RootCircuit::build(Self::build_streams)?;

        Ok(Self {
            circuit,
            position_in: handles.position_in,
            velocity_in: handles.velocity_in,
            force_in: handles.force_in,
            target_in: handles.target_in,
            block_in: handles.block_in,
            block_slope_in: handles.block_slope_in,
            new_position_out: handles.new_position_out,
            new_velocity_out: handles.new_velocity_out,
            highest_block_out: handles.highest_block_out,
            floor_height_out: handles.floor_height_out,
            position_floor_out: handles.position_floor_out,
        })
    }

    pub fn step(&mut self) -> Result<(), dbsp::Error> {
        self.circuit.step()
    }

    fn build_streams(circuit: &mut RootCircuit) -> Result<BuildHandles, AnyError> {
        let (positions, position_in) = circuit.add_input_zset::<Position>();
        let (velocities, velocity_in) = circuit.add_input_zset::<Velocity>();
        let (forces, force_in) = circuit.add_input_zset::<Force>();
        let (targets, target_in) = circuit.add_input_zset::<Target>();
        let (blocks, block_in) = circuit.add_input_zset::<Block>();
        let (slopes, block_slope_in) = circuit.add_input_zset::<BlockSlope>();

        let highest_pair = highest_block_pair(&blocks);
        let highest = highest_pair.map(|(hb, _)| *hb);
        let floor_height = floor_height_stream(&highest_pair, &slopes);

        let pos_floor = position_floor_stream(&positions, &floor_height);

        fn within_grace(pf: &PositionFloor) -> bool {
            pf.position.z.into_inner() <= pf.z_floor.into_inner() + GRACE_DISTANCE
        }
        let unsupported = pos_floor.filter(|pf| !within_grace(pf));
        let standing = pos_floor.filter(within_grace);

        let unsupported_positions = unsupported.map(|pf| pf.position);
        let all_new_vel = new_velocity_stream(&velocities, &forces);
        let unsupported_velocities = all_new_vel.map_index(|v| (v.entity, *v)).join(
            &unsupported.map_index(|pf| (pf.position.entity, ())),
            |_, vel, _| *vel,
        );
        let new_pos_unsupported =
            new_position_stream(&unsupported_positions, &unsupported_velocities);

        let (new_pos_standing, new_vel_standing) =
            standing_motion_stream(&standing, &floor_height, &all_new_vel);

        let base_pos = new_pos_unsupported.plus(&new_pos_standing);
        let new_vel = unsupported_velocities.plus(&new_vel_standing);

        let fear = fear_level_stream(&positions);
        let decisions = movement_decision_stream(&fear, &targets, &positions);

        let moved_pos = apply_movement(&base_pos, &decisions);

        Ok(BuildHandles {
            position_in,
            velocity_in,
            force_in,
            target_in,
            block_in,
            block_slope_in,
            new_position_out: moved_pos.output(),
            new_velocity_out: new_vel.output(),
            highest_block_out: highest.output(),
            floor_height_out: floor_height.output(),
            position_floor_out: pos_floor.output(),
        })
    }

    /// Returns a reference to the input handle for feeding position records into the circuit.
    ///
    /// Use this handle to provide entity position data for processing by the dataflow circuit.
    ///
    /// # Examples
    ///
    /// ```rust,no_run
    /// # use lille::prelude::*;
    /// let circuit = DbspCircuit::new().expect("circuit construction failed");
    /// let position_handle = circuit.position_in();
    /// // Feed positions into the circuit using `position_handle`
    /// ```
    pub fn position_in(&self) -> &ZSetHandle<Position> {
        &self.position_in
    }

    /// Returns a reference to the input handle for feeding velocity records into the circuit.
    ///
    /// Use this handle to provide entity velocity data for each simulation step.
    ///
    /// # Examples
    ///
    /// ```rust,no_run
    /// # use lille::prelude::*;
    /// let circuit = DbspCircuit::new().expect("circuit construction failed");
    /// let velocity_in = circuit.velocity_in();
    /// velocity_in.push(
    ///     Velocity {
    ///         entity: 1,
    ///         vx: OrderedFloat(0.0),
    ///         vy: OrderedFloat(0.0),
    ///         vz: OrderedFloat(0.0),
    ///     },
    ///     1,
    /// );
    /// ```
    pub fn velocity_in(&self) -> &ZSetHandle<Velocity> {
        &self.velocity_in
    }

    /// Returns a reference to the input handle for feeding force records into the circuit.
    ///
    /// Use this handle to supply external forces acting on entities. If a
    /// force is omitted for an entity, only gravity is applied.
    ///
    /// # Examples
    ///
    /// ```rust,no_run
    /// # use lille::prelude::*;
    /// # use ordered_float::OrderedFloat;
    /// let circuit = DbspCircuit::new().expect("circuit construction failed");
    /// let force_in = circuit.force_in();
    /// force_in.push(
    ///     Force {
    ///         entity: 1,
    ///         fx: OrderedFloat(5.0),
    ///         fy: OrderedFloat(0.0),
    ///         fz: OrderedFloat(0.0),
    ///         mass: Some(OrderedFloat(5.0)),
    ///     },
    ///     1,
    /// );
    /// ```
    pub fn force_in(&self) -> &ZSetHandle<Force> {
        &self.force_in
    }

    /// Returns a reference to the input handle for entity targets.
    ///
    /// # Examples
    ///
    /// ```rust,no_run
    /// # use ordered_float::OrderedFloat;
    /// # use lille::dbsp_circuit::{DbspCircuit, Target as DbspTarget};
    /// let circuit = DbspCircuit::new().expect("circuit construction failed");
    /// let handle = circuit.target_in();
    /// handle.push(
    ///     DbspTarget { entity: 1, x: OrderedFloat(10.0), y: OrderedFloat(5.0) },
    ///     1,
    /// );
    /// ```
    pub fn target_in(&self) -> &ZSetHandle<Target> {
        &self.target_in
    }

    /// Returns a reference to the input handle for feeding block records into the circuit.
    ///
    /// Use this handle to provide block data as input for each computation step.
    ///
    /// # Examples
    ///
    /// ```rust,no_run
    /// # use lille::prelude::*;
    /// let circuit = DbspCircuit::new().expect("circuit construction failed");
    /// let block_handle = circuit.block_in();
    /// // Feed block data into the circuit using `block_handle`
    /// ```
    pub fn block_in(&self) -> &ZSetHandle<Block> {
        &self.block_in
    }

    /// Returns a reference to the input handle for feeding block slope records into the circuit.
    ///
    /// Use this handle to supply slope gradient data for blocks, enabling
    /// slope-aware floor height calculations.
    ///
    /// # Examples
    ///
    /// ```rust,no_run
    /// # use lille::prelude::*;
    /// let circuit = DbspCircuit::new().expect("circuit construction failed");
    /// let slope_handle = circuit.block_slope_in();
    /// // Feed block slope data into the circuit using `slope_handle`
    /// ```
    pub fn block_slope_in(&self) -> &ZSetHandle<BlockSlope> {
        &self.block_slope_in
    }

    /// Returns a reference to the output handle for newly computed entity positions.
    ///
    /// The output handle provides access to the set of updated positions after each circuit step.
    ///
    /// # Examples
    ///
    /// ```rust,no_run
    /// # use lille::prelude::*;
    /// let circuit = DbspCircuit::new().expect("circuit construction failed");
    /// let new_positions = circuit.new_position_out();
    /// // Read new positions from the output handle
    /// ```
    pub fn new_position_out(&self) -> &OutputHandle<OrdZSet<NewPosition>> {
        &self.new_position_out
    }

    /// Returns a reference to the output handle for newly computed velocities.
    ///
    /// The output contains updated velocity records for all entities after applying
    /// the circuit's physics computations, such as gravity.
    ///
    /// # Examples
    ///
    /// ```rust,no_run
    /// # use lille::prelude::*;
    /// let circuit = DbspCircuit::new().expect("circuit construction failed");
    /// let velocities = circuit.new_velocity_out();
    /// // Use `velocities` to read updated velocity data.
    /// ```
    pub fn new_velocity_out(&self) -> &OutputHandle<OrdZSet<NewVelocity>> {
        &self.new_velocity_out
    }

    /// Returns a reference to the output handle for the highest block at each (x, y) coordinate.
    ///
    /// The output contains `HighestBlockAt` records representing the maximum `z` value for each `(x, y)`
    /// position in the input block data.
    ///
    /// # Examples
    ///
    /// ```rust,no_run
    /// # use lille::prelude::*;
    /// let circuit = DbspCircuit::new().expect("circuit construction failed");
    /// let highest_block_handle = circuit.highest_block_out();
    /// // Use `highest_block_handle` to read aggregated highest block data.
    /// ```
    pub fn highest_block_out(&self) -> &OutputHandle<OrdZSet<HighestBlockAt>> {
        &self.highest_block_out
    }

    /// Returns a reference to the output handle for calculated floor heights.
    ///
    /// The output contains `FloorHeightAt` records representing the computed
    /// floor height at each `(x, y)` position, incorporating block heights and
    /// optional slope gradients.
    ///
    /// # Examples
    ///
    /// ```rust,no_run
    /// # use lille::prelude::*;
    /// let circuit = DbspCircuit::new().expect("circuit construction failed");
    /// let floor_heights = circuit.floor_height_out();
    /// // Read computed floor heights from the output handle
    /// ```
    pub fn floor_height_out(&self) -> &OutputHandle<OrdZSet<FloorHeightAt>> {
        &self.floor_height_out
    }

    /// Returns a reference to the output handle for entity positions joined with
    /// floor height.
    ///
    /// The output contains [`PositionFloor`] records that pair each entity's
    /// [`Position`] with the discrete [`FloorHeightAt`] value at its grid cell.
    /// Use this handle to read the results of the position-to-floor join after
    /// each call to [`DbspCircuit::step`].
    ///
    /// # Examples
    ///
    /// ```rust,no_run
    /// # use lille::prelude::*;
    /// let mut circuit = DbspCircuit::new().expect("circuit construction failed");
    /// let joined = circuit.position_floor_out();
    /// // Read joined records from `joined`
    /// ```
    pub fn position_floor_out(&self) -> &OutputHandle<OrdZSet<PositionFloor>> {
        &self.position_floor_out
    }

    /// Clears all input collections to remove accumulated records.
    ///
    /// Input ZSets retain data across [`DbspCircuit::step`] calls. Invoke this method after
    /// processing outputs each frame to ensure that stale input data does not affect future
    /// computations.
    ///
    /// # Examples
    ///
    /// ```rust,no_run
    /// # use lille::prelude::*;
    /// let mut circuit = DbspCircuit::new().expect("circuit construction failed");
    /// circuit.clear_inputs();
    /// ```
    pub fn clear_inputs(&mut self) {
        self.position_in.clear_input();
        self.velocity_in.clear_input();
        self.force_in.clear_input();
        self.target_in.clear_input();
        self.block_in.clear_input();
        self.block_slope_in.clear_input();
    }
}

#[cfg(test)]
mod tests;
