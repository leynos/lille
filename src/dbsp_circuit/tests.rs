//! Tests for DBSP circuit behaviour, including grace distance and health aggregation.
use super::*;
use crate::dbsp_circuit::streams::health_delta_stream;
use crate::dbsp_circuit::{DamageEvent, DamageSource, HealthDelta, HealthState};
use crate::GRACE_DISTANCE;
use dbsp::RootCircuit;
use rstest::rstest;

fn make_pf(z: f64, z_floor: f64) -> PositionFloor {
    PositionFloor {
        position: Position {
            entity: 1,
            x: 0.0.into(),
            y: 0.0.into(),
            z: z.into(),
        },
        z_floor: z_floor.into(),
    }
}

#[rstest]
#[case(10.0, 10.0)]
#[case(10.05, 10.0)]
#[case(10.1, 10.0)]
#[case(0.0, 0.0)]
#[case(-10.0, -10.0)]
#[case(-10.05, -10.0)]
#[case(f64::NAN, 10.0)]
#[case(10.0, f64::NAN)]
#[case(f64::NAN, f64::NAN)]
fn within_grace(#[case] z: f64, #[case] z_floor: f64) {
    let pf = make_pf(z, z_floor);
    if z.is_nan() || z_floor.is_nan() {
        // Comparisons with NaN are always false -> outside grace.
        let within = pf.position.z.into_inner() <= pf.z_floor.into_inner() + GRACE_DISTANCE;
        assert!(!within);
    } else {
        assert!(pf.position.z.into_inner() <= pf.z_floor.into_inner() + GRACE_DISTANCE);
    }
}

#[rstest]
#[case(11.0, 10.0)]
#[case(10.0 + GRACE_DISTANCE, 10.0)]
fn beyond_grace_or_at_boundary(#[case] z: f64, #[case] z_floor: f64) {
    let pf = make_pf(z, z_floor);
    if pf.position.z.into_inner() == pf.z_floor.into_inner() + GRACE_DISTANCE {
        assert!(pf.position.z.into_inner() <= pf.z_floor.into_inner() + GRACE_DISTANCE);
    } else {
        assert!(pf.position.z.into_inner() > pf.z_floor.into_inner() + GRACE_DISTANCE);
    }
}

fn run_health_delta(health: HealthState, events: &[(DamageEvent, i32)]) -> Vec<HealthDelta> {
    let (circuit, (health_handle, damage_handle, output)) = RootCircuit::build(|circuit| {
        let (health_stream, health_handle) = circuit.add_input_zset::<HealthState>();
        let (damage_stream, damage_handle) = circuit.add_input_zset::<DamageEvent>();
        let output = health_delta_stream(&health_stream, &damage_stream).output();
        Ok((health_handle, damage_handle, output))
    })
    .expect("failed to build health circuit");

    health_handle.push(health, 1);
    for (event, weight) in events {
        damage_handle.push(*event, i64::from(*weight));
    }

    circuit.step().expect("health circuit step failed");
    output
        .consolidate()
        .iter()
        .map(|(delta, _, _)| delta)
        .collect()
}

fn assert_health_delta_test(
    health_state: HealthState,
    events: &[(DamageEvent, i32)],
    expected_delta: i32,
    expected_death: bool,
    expected_seq: Option<u32>,
) {
    let deltas = run_health_delta(health_state, events);
    assert_eq!(deltas.len(), 1);
    let delta = deltas[0];
    assert_eq!(delta.delta, expected_delta);
    assert_eq!(delta.death, expected_death);
    assert_eq!(delta.seq, expected_seq);
}

fn run_health_delta_test(
    entity: u64,
    current: u16,
    max: u16,
    events: Vec<(u16, DamageSource, u64, Option<u32>)>,
    expected_delta: i32,
    expected_death: bool,
    expected_seq: Option<u32>,
) {
    let health = HealthState {
        entity,
        current,
        max,
    };
    let damage_events: Vec<(DamageEvent, i32)> = events
        .into_iter()
        .map(|(amount, source, at_tick, seq)| {
            (
                DamageEvent {
                    entity,
                    amount,
                    source,
                    at_tick,
                    seq,
                },
                1,
            )
        })
        .collect();

    assert_health_delta_test(
        health,
        &damage_events,
        expected_delta,
        expected_death,
        expected_seq,
    );
}

#[rstest]
#[case(80, 100, 50, 20)]
#[case(10, 50, 5, 5)]
fn healing_clamped_to_max(
    #[case] current: u16,
    #[case] max: u16,
    #[case] heal: u16,
    #[case] expected_delta: i32,
) {
    let health = HealthState {
        entity: 1,
        current,
        max,
    };
    let event = DamageEvent {
        entity: 1,
        amount: heal,
        source: DamageSource::Script,
        at_tick: 5,
        seq: Some(1),
    };

    let deltas = run_health_delta(health, &[(event, 1)]);
    assert_eq!(deltas.len(), 1);
    let delta = deltas[0];
    assert_eq!(delta.delta, expected_delta);
    assert!(!delta.death);
    assert_eq!(delta.seq, Some(1));
}

#[rstest]
#[case(Some(3))]
#[case(None)]
fn duplicate_damage_events_idempotent(#[case] seq: Option<u32>) {
    let health = HealthState {
        entity: 2,
        current: 90,
        max: 100,
    };
    let event = DamageEvent {
        entity: 2,
        amount: 30,
        source: DamageSource::External,
        at_tick: 9,
        seq,
    };

    assert_health_delta_test(health, &[(event, 1), (event, 1)], -30, false, seq);
}

#[rstest]
fn sequenced_events_with_same_seq_in_same_tick_are_deduplicated() {
    let health = HealthState {
        entity: 7,
        current: 70,
        max: 100,
    };
    let first = DamageEvent {
        entity: 7,
        amount: 20,
        source: DamageSource::External,
        at_tick: 8,
        seq: Some(11),
    };
    // Provide the duplicate event with an identical payload to mirror the ingress
    // first-write-wins policy: later `(entity, tick, seq)` writes are ignored, and
    // the matching payload ensures the circuit's debug assertions are satisfied.
    let second = DamageEvent {
        entity: 7,
        amount: 20,
        source: DamageSource::External,
        at_tick: 8,
        seq: Some(11),
    };

    assert_health_delta_test(health, &[(first, 1), (second, 1)], -20, false, Some(11));
}

#[rstest]
#[case::unsequenced_distinct_sources(6, 40, 100, vec![(15, DamageSource::External, 4, None), (25, DamageSource::Script, 4, None)], 10, false, None)]
#[case::unsequenced_duplicate_payloads_filtered(6, 40, 100, vec![(15, DamageSource::External, 4, None), (15, DamageSource::External, 4, None)], -15, false, None)]
#[case::multiple_events_max_seq(5, 100, 120, vec![(60, DamageSource::External, 10, Some(1)), (20, DamageSource::Script, 10, Some(4))], -40, false, Some(4))]
#[case::healing_from_zero(4, 0, 80, vec![(30, DamageSource::Script, 3, None)], 30, false, None)]
#[case::over_healing_clamped(5, 0, 80, vec![(150, DamageSource::Script, 4, None)], 80, false, None)]
fn health_delta_scenarios(
    #[case] entity: u64,
    #[case] current: u16,
    #[case] max: u16,
    #[case] events: Vec<(u16, DamageSource, u64, Option<u32>)>,
    #[case] expected_delta: i32,
    #[case] expected_death: bool,
    #[case] expected_seq: Option<u32>,
) {
    run_health_delta_test(
        entity,
        current,
        max,
        events,
        expected_delta,
        expected_death,
        expected_seq,
    );
}

#[rstest]
fn lethal_damage_sets_death_flag() {
    run_health_delta_test(
        3,
        20,
        50,
        vec![(40, DamageSource::External, 2, Some(7))],
        -20,
        true,
        Some(7),
    );
}
