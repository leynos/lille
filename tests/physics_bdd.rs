//! Physics behaviour-driven development tests using the DBSP circuit.
//!
//! This module tests physics rules such as gravity effects on entity positions
//! through the declarative dataflow circuit.

use lille::dbsp_circuit::{DbspCircuit, NewPosition, Position};

#[test]
fn entity_falls_due_to_gravity() {
    let mut circuit = DbspCircuit::new().expect("Failed to create DBSP circuit");

    circuit.position_in().push(
        Position {
            entity: 1,
            x: 0.0.into(),
            y: 0.0.into(),
            z: 1.0.into(),
        },
        1,
    );
    circuit.step().expect("Failed to step DBSP circuit");

    let output = circuit.new_position_out().consolidate();
    let results: Vec<NewPosition> = output.iter().map(|(p, _, _)| p.clone()).collect();
    assert_eq!(results.len(), 1);
    assert_eq!(results[0].entity, 1);
    assert_eq!(results[0].z.into_inner(), 1.0 + lille::GRAVITY_PULL);
}
