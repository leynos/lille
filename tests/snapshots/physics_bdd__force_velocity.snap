---
source: tests/physics_bdd.rs
assertion_line: 95
expression: "&ddlog.velocity_deltas"
---
[
  NewVelocity(
    entity: 1,
    vx: 0.1,
    vy: 0.0,
    vz: 0.0,
  ),
]
